README.md
# Braincube CSR Client

This is the Braincube CSR client repository.

The goal of this repository is to provide a CLI client to perform CSR request and onboard client onto Data Repository.

This CLI client asks some questions to do the operations. No options are needed.

## Installation (for braincube usage only)
To install, just run :
`npm i -g @braincube/braincube-csr-client`

This will install this package as a global package, and now will be available as the command `braincube-csr-internal`.

## Installation (from github public repository)
Just install all dependencies by typing :
```
    npm install
```

## Build
For build the cjs and mjs code type the following command :
```
    npm run build
```
For building the binary for the supported os platforms type
```
    npm run build-bin
```

## Run

* From Braincube installation use
    ```
        braincube-csr-internal or npx braincube-csr-internal
    ```
    commands, to display help/usage.

* From scratch building (like public repository) go to `./.bin` and type the binary of your plateform choice like that :
    ```
        ./braincube-csr-client-linux
    ```

## Usage

###### Usage: braincube-csr-internal [options] [command]

  A CLI client for onboarding onto Data Repository, on check connection

  Options:

    -V, --version  output the version number
    -h, --help     output usage information

  Commands:

    onboard|o  ${optionnal_private_key_pem_path}     Onboard onto a Data Repository with the possibility 
                                                     to import your own RSA privateKey (4096 bits recommended)
    check|c                                          Check connection to a Data Repository



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), 
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.3] - 15-01-2021

### Fixed

* Fix cdn directory deployment

## [0.1.2] - 15-01-2021

### Internal

* Add internal improvement ci

## [0.1.1] - 14-01-2021

### Added

* Possibility to import your own RSA private key to perform CSR request

## [0.1.0] - 26-07-2019

### Added

* updated dependencies
* better CI to allow a proper `npm pack`

### Fixed

* Windows build is now working again (broken with 0.0.9)

## [0.0.9] - 24-07-2019

### Added

* The domain is now a text field : with no implied .mybraincube.com

## [0.0.8] - 27-06-2019

### Added

* Add comments for open source code

### Fixed

* Version option result are wrong

## [0.0.7] - 24-04-2019

### Added

* The domain is now a text field, allowing to work on other domains

## [0.0.6] - 14-02-2019

### Added

* Increase polling timeout to 10 min

## [0.0.5] - 16-08-2018

### Fixed

* Miss default sso instance

## [0.0.4] - 16-08-2018

### Fixed

* Add ci tests

## [0.0.3] - 16-08-2018

### Added

* Add binary build step on ci

## [0.0.2] - 28-06-2018

### Added

* Add mjs generation for internal script use

## [0.0.1] - 19-06-2018

## Added

* This is the first and only version which contains all awesomes features to generate a CSR request, and onboard a client onto a braincube data repository



tests/check.test.js
const check = require('../lib/check');
const should = require("should");

describe('check', () => {

    it.skip ('test', () => {
        //test to be perform ;)
    });



})

tests/onboard.test.js
// eslint-disable-next-line import/no-unresolved,no-unused-vars
const should = require('should');
const onboard = require('../lib/onboard.mjs');

describe('/lib/onboard', () => {

    it.skip('post board ok', (done) => {
        // test to be perform ;) fucking jest and mjs
        onboard.postBoard('braincubetransfer.test.mybraincube.com').then((server) => {
            done(server);
        });
    });
});


lib/cli.mjs

import program from 'commander';

import * as onboard from './onboard.mjs';
import * as check from './check.mjs';

const softwareVersion = '<@VERSION@>';

// Version command option
program
    .version(softwareVersion)
    .description('A CLI client for onboarding on Braincube servers and checking connection');

// Onboard command option
program
    .command('onboard')
    .description('Onboard to braincube')
    .alias('o')
    .action(() => onboard.cli(process.argv.slice(2)[1]));

// Check file uploading test
program
    .command('check')
    .description('Check connection to Braincube (need to be onboarded first)')
    .alias('c')
    .action(check.cli);

// Help command option (also default choice)
program
    .on('--help', () => {
        /* eslint-disable no-console */
        console.log('\n  Examples:');
        console.log('\nto onboard :');
        console.log('\t$ ./braincube-csr-client-linux o');
        console.log('\t$ ./braincube-csr-client-linux onboard');
        console.log('to onboard using your own RSA private key (4096 bits recommended):');
        console.log('\t$ ./braincube-csr-client-linux onboard <private_key_pem_path>');
        console.log('to check :');
        console.log('\t$ ./braincube-csr-client-linux c');
        console.log('\t$ ./braincube-csr-client-linux check');
        console.log('');
    });

// Parse the option args to inquirer
program
    .parse(process.argv);

// Check if there are no args go to help !
if (!process.argv.slice(2).length) {
    program.help();
}


lib/check.mjs
import inquirer from 'inquirer';
import fs from 'fs';
import path from 'path';
import forge from 'node-forge';
import readline from 'readline';
import util from 'util';

import { checkConnection } from './checkConn.mjs';

const WAIT_PUSH = 60000;


/**
 * Log a message into the stdout
 * @param str
 * @param newLine
 */
function log(str, newLine = true) {
    if (newLine) {
        process.stdout.write('\n');
    } else {
        readline.clearLine(process.stdout, 0);
        readline.cursorTo(process.stdout, 0);
    }
    process.stdout.write(str);
}

/**
 * Function to launch cli inquiring
 * @returns {Promise<void>}
 */
export async function cli() {
    inquirer
        .prompt([
            { // braincube transfer host. PROD or DEV
                type: 'list',
                name: 'server',
                message: 'SSO intance',
                choices: ['braincubetransfer', 'braincubetransfer.test'],
                default: 0,
                filter: (input) => `${input}.mybraincube.com`
            },
            {
                type: 'list',
                name: 'onboard',
                message: 'Onboard file',
                choices: () => {
                    return fs.readdirSync(path.resolve(process.cwd()))
                        .filter((p) => p.indexOf('.onboard') >= 0) // filter only onboard files
                        .map((p) => path.relative(process.cwd(), p));
                }
            }
        ])
        .then(async (answers) => {
            /* eslint-disable no-console */
            console.log(fs.readFileSync(path.resolve(process.cwd(), answers.onboard)).toString());
            const onboardInfo = JSON.parse(fs.readFileSync(path.resolve(process.cwd(), answers.onboard)).toString());
            log(`Found onboard info ${util.inspect(onboardInfo)}`);
            const cert = forge.pki.certificateFromPem(fs.readFileSync(path.resolve(process.cwd(), onboardInfo.cert)));

            const key = forge.pki.privateKeyFromPem(fs.readFileSync(path.resolve(process.cwd(), onboardInfo.privateKey)));
            log(`Checking dataRepository ${onboardInfo.productId}`);
            try {
                await checkConnection(answers.server,
                    onboardInfo.productId,
                    forge.pki.certificateToPem(cert),
                    forge.pki.privateKeyToPem(key),
                    WAIT_PUSH);
            } catch (e) {
                /* eslint-disable no-console */
                console.error(e);
            }
        });
}

export default {};


lib/checkConn.mjs
import https from 'https';
import tls from 'tls';
import readline from 'readline';

/**
 * Log a message into the stdout
 * @param str
 * @param newLine
 */
function log(str, newLine = true) {
    if (newLine) {
        process.stdout.write('\n');
    } else {
        readline.clearLine(process.stdout, 0);
        readline.cursorTo(process.stdout, 0);
    }
    process.stdout.write(str);
}

/**
 * POST a request using the given options
 * @param httpsOptions Object https options
 * @return {Promise<any>}
 */
async function postFile({ httpsOptions }) {
    return new Promise((resolve, reject) => {
        try {
            let body = '';
            https.request(httpsOptions,
                (resp) => {
                    const { statusCode } = resp;
                    resp.on('data', (chunk) => {
                        body += chunk;
                    });
                    resp.on('end', () => {
                        resolve({ statusCode, body });
                    });
                }).on('error', (err) => {
                reject(new Error(`An error occured during POST boarding information${err}`));
            }).end(' ');
        } catch (e) {
            reject(e);
        }
    });
}

/**
 * Try to POST a trash file, and waits for 200 HTTP code
 * @param server braincubetransfer hostname
 * @param productId Data Repository productId
 * @param certPem certificate in PEM format
 * @param privKeyPem private key certificate
 * @param timeout global timeout
 * @return {Promise<{ statusCode }>}
 */
export async function checkConnection(server, productId, certPem, privKeyPem, timeout) {
    log(`Check connection on ${server} for ${productId}\n`);
    const tlsOptions = {
        cert: certPem,
        key: privKeyPem
    };
    const tlsContext = tls.createSecureContext(tlsOptions);
    const httpsOptions = {
        protocol: 'https:',
        host: server,
        path: `/upload/${productId}`,
        method: 'POST',
        headers: {
            'File-Name': 'test.trash',
            'content-length': '1'
        },
        secureContext: tlsContext
    };
    httpsOptions.agent = https.Agent(httpsOptions);
    const sleepTime = 2000;
    const validReturnCode = 200;
    let tryCount = 0;
    return new Promise((resolve, reject) => {
        let postTimeout;
        let rejTimeout;
        const repeat = async () => {
            let status;
            try {
                const { statusCode } = await postFile({ httpsOptions });
                status = statusCode;
            } catch (err) {
                status = 500;
            }
            log(`post file result was ${status} ${'.'.repeat(tryCount)}`, false);
            if (status !== validReturnCode) {
                postTimeout = setTimeout(repeat, sleepTime);
            } else {
                log('Connection SUCCESSFULL');
                clearTimeout(rejTimeout);
                resolve({ status });
            }
            tryCount++;
        };
        repeat();
        rejTimeout = setTimeout(() => {
            clearTimeout(postTimeout);
            reject(new Error(`Status code ${validReturnCode} not found in ${timeout} ms`));
        }, timeout);
    });
}

export default {};


lib/onboard.mjs
/* eslint-disable max-nested-callbacks */
import inquirer from 'inquirer';
import fs from 'fs';
import path from 'path';
import forge from 'node-forge';
import https from 'https';
import url from 'url';
import opn from 'opn';
import readline from 'readline';

import { checkConnection } from './checkConn.mjs';

const onBoardingPath = '/sso-server/ws/transfer/board';

const WAIT_POLL = 600000; // 10 min
const WAIT_PUSH = 60000;

/**
 * Log a message into the stdout
 * @param str
 * @param newLine
 */
function log(str, newLine = true) {
    if (typeof str !== 'string') {
        process.stdout.write('\nError to display log ' + typeof str);
        return;
    }
    if (newLine) {
        process.stdout.write('\n');
    } else {
        readline.clearLine(process.stdout, 0);
        readline.cursorTo(process.stdout, 0);
    }
    process.stdout.write(str);
}

/**
 * Post the CSR to /sso-server/ws/transfer/sign?code=${code}
 * @param server braincubetransfer host
 * @param csr csr pem content
 * @param code request code
 * @return {Promise<any>}
 */
export async function postCsr({ server, csr, code }) {
    return new Promise((resolve, reject) => {
        // TEMPCode : envoyer csr dans le body en url
        // POST -> https://mybraincube.com
        const req = https.request({
            protocol: 'https:',
            host: server,
            path: `/sso-server/ws/transfer/sign?code=${code}`,
            method: 'POST'
        }, (res) => {
            let data = '';
            const { statusCode } = res;
            res.on('data', (chunk) => {
                data += chunk;
            });
            res.on('end', () => {
                if (statusCode === 200) {
                    resolve({ statusCode, data });
                } else {
                    reject(new Error(`Error while posting csr. Bad status code : ${statusCode}. ${data}`));
                }
            });
        });
        req.on('error', (err) => {
            reject(new Error(`An error occured during POST boarding information${err}`));
        });
        req.end(csr);
    });
}


/**
 * Make a GET call to the location, and resolve body and statusCode
 * @param location
 * @return {Promise<{ statusCode, body }>}
 */
async function pollOnce({ location }) {
    return new Promise((resolve, reject) => {
        let body = '';
        https.get(location, (res) => {
            const { statusCode } = res;
            res.on('data', (chunk) => { body += chunk; });
            res.on('end', () => {
                resolve({ statusCode, body });
            });
        }).on('error', (e) => {
            reject(new Error(`Error during polling ${location} : ${e}`));
        });
    });
}


/**
 * Poll the url using GET method waiting for HTTP 200 code
 * @param location url to poll
 * @param timeout global timeout
 * @return {Promise<{ statusCode, body }>}
 */
export async function poll({ location, timeout }) {
    // 206 -> en cours
    // 200-> ok
    let count = timeout;
    const sleepTime = 2000;
    const validReturnCode = 200;
    return new Promise((resolve, reject) => {
        let readTimeout;
        let rejTimeout;
        const repeat = async () => {
            const { statusCode, body } = await pollOnce({ location });
            log(`poll result was ${statusCode}. you have ${count / 1000} s`, false);
            if (statusCode !== validReturnCode) {
                readTimeout = setTimeout(repeat, sleepTime);
            } else {
                clearTimeout(rejTimeout);
                resolve({ statusCode, body });
            }
            count -= sleepTime;
        };

        repeat().catch((err) => {
            readTimeout = setTimeout(repeat, sleepTime);
            reject(err);
        });

        rejTimeout = setTimeout(() => {
            clearTimeout(readTimeout);
            reject(new Error(`Status code ${validReturnCode} not found in ${timeout} ms`));
        }, timeout);
    });
}

/**
 * Make a POST request to the boarding url, and resolve with the location (from headers) and the clientUrl (in reponse's body)
 * @param server hostname
 * @return {Promise<{ location, clientUrl }>}
 */
export async function postBoard({ server }) {
    log(`Onboarding to ${server}`);
    return new Promise((resolve, reject) => {
        let clientUrl = '';
        https.request({
            protocol: 'https:',
            host: server,
            path: onBoardingPath,
            method: 'POST'
        }, (resp) => {
            const location = resp.headers.location;
            resp.on('data', (chunk) => {
                clientUrl += chunk;
            });
            resp.on('end', () => {
                resolve({ location, clientUrl });
            });
        }).on('error', (err) => {
            reject(new Error(`An error occured during POST boarding information${err}`));
        }).end();
    });
}

/**
 * Genetate RSA key pair.
 * @param externalPrivKeyPath
 * @param [bits] the size for the private key in bits, defaults to 4096.
 * @return {Promise<keys>}
 */
export async function generateKeyPair(externalPrivKeyPath, bits = 4096) {

    if (externalPrivKeyPath) {
        log('Generating Key Pair using external RSA private key from : ' + externalPrivKeyPath);
        try {
            const privKey = fs.readFileSync(externalPrivKeyPath);
            const privateKey = forge.pki.privateKeyFromPem(privKey);
            const publicKey = forge.pki.setRsaPublicKey(privateKey.n, privateKey.e);
            const keys = {
                privateKey: privateKey,
                publicKey: publicKey
            };
            return Promise.resolve(keys);
        } catch (e) {
            throw new Error(
                'File can\'t be read or your private key format isn\'t good '
                + '(must be an RSA format with a recommended length of 4096 bits)'
            );
        }
    }

    const keys = forge.pki.rsa.generateKeyPair(bits);
    return Promise.resolve(keys);
}

/**
 * Save key pair into 2 files, and resolve once it's done
 * @param paths Object
 * @param paths.publicKey public key path
 * @param paths.privateKey private key path
 * @param keys Object
 * @param keys.publicKey
 * @param keys.privateKey
 * @return {Promise<>}
 */
export async function saveKeyPair(paths, keys) {
    return new Promise((resolve, reject) => {
        try {
            fs.writeFileSync(paths.publicKey, forge.ssh.publicKeyToOpenSSH(keys.publicKey));
            fs.writeFileSync(paths.privateKey, forge.pki.privateKeyToPem(keys.privateKey), { mode: 0o600 });
            resolve();
        } catch (e) {
            reject(e);
        }
    });
}

/**
 * Save certificate and resolves when it's done
 * @param paths Object
 * @param paths.cert cert path
 * @param cert cert in pem format
 * @return {Promise<any>}
 */
export async function saveCert(paths, cert) {
    return new Promise((resolve, reject) => {
        try {
            const forgeCert = forge.pki.certificateFromPem(cert);
            fs.writeFileSync(paths.cert, forge.pki.certificateToPem(forgeCert));
            resolve();
        } catch (e) {
            reject(e);
        }
    });
}
async function saveOnboardFile(paths, onboardInfo) {
    return new Promise((resolve, reject) => {
        try {
            fs.writeFileSync(paths.onboard, JSON.stringify(onboardInfo));
            resolve();
        } catch (e) {
            reject(e);
        }
    });
}

/**
 * Save CSR into file and resolve when its done.
 * @param paths
 * @param paths.csr csr path
 * @param pemCsr csr in PEM format
 * @return {Promise<any>}
 */
export async function saveCsr(paths, pemCsr) {
    return new Promise((resolve, reject) => {
        try {
            fs.writeFileSync(paths.csr, pemCsr);
            resolve();
        } catch (e) {
            reject(e);
        }
    });
}

/**
 * Generate CSR, and set user mail, productid and pulbic key.
 * @param userInfo
 * @param userInfo.user user mail
 * @param userInfo.productId wanted productId
 * @param keys
 * @param keys.publicKey public key
 * @return {Promise<csr>} forge.pki csr object
 */
export async function getCsr(userInfo, keys) {
    return new Promise((resolve, reject) => {
        try {
            const csr = forge.pki.createCertificationRequest();
            csr.publicKey = keys.publicKey;
            csr.setSubject([
                {
                    shortName: 'CN',
                    value: userInfo.user
                },
                {
                    name: 'UniqueIdentifier',
                    type: '2.5.4.45', // This is the OID for `UniqueIdentifier`,Cf http://www.alvestrand.no/objectid/2.5.4.45.html
                    value: userInfo.productId
                }
            ]);
            resolve(csr);
        } catch (e) {
            reject(e);
        }
    });
}


/**
 * Function to launch cli inquiring
 * @param externalPrivKeyPath
 * @returns {Promise<void>}
 */
export async function cli(externalPrivKeyPath) {
    /**
     * Prompt using inquirer
     */
    inquirer.prompt([
        { // braincube transfer host. PROD or DEV
            type: 'input',
            name: 'server',
            message: 'SSO instance (braincubetransfer.mybraincube.com)',
            default: 'braincubetransfer.mybraincube.com'
        },
        { // private key filename
            type: 'input',
            name: 'privKeyPath',
            message: 'Generated RSA private key file name (public key file name will be <private>.pub)',
            default: 'id_rsa'
        },
        { // should we send a trash file ?
            type: 'confirm',
            name: 'check',
            default: true,
            message: 'Check connection by sending .trash file (this will start the docker container) ?'
        },
        { // should we overwrite existing key/cert/pem ?
            type: 'confirm',
            name: 'overwrite',
            default: false,
            // eslint-disable-next-line max-len
            message: (answ) => `${answ.privKeyPath} or ${answ.privKeyPath}.pub  or ${answ.privKeyPath}.pem  or ${answ.privKeyPath}.csr already exists. Overwrite ?`,
            when: (answ) => {
                return fs.existsSync(path.resolve(process.cwd(), answ.privKeyPath))
                    || fs.existsSync(path.resolve(process.cwd(), `${answ.privKeyPath}.pub`))
                    || fs.existsSync(path.resolve(process.cwd(), `${answ.privKeyPath}.pem`))
                    || fs.existsSync(path.resolve(process.cwd(), `${answ.privKeyPath}.csr`))
                    || fs.existsSync(path.resolve(process.cwd(), `${answ.privKeyPath}.onboard`));
            }
        }
    ]).then(async (answers) => {
        if (answers.overwrite !== false) {
            // first generate CSR
            try {
                const paths = {
                    privateKey: path.resolve(process.cwd(), `${answers.privKeyPath}`),
                    publicKey: path.resolve(process.cwd(), `${answers.privKeyPath}.pub`),
                    csr: path.resolve(process.cwd(), `${answers.privKeyPath}.csr`),
                    cert: path.resolve(process.cwd(), `${answers.privKeyPath}.pem`),
                    onboard: path.resolve(process.cwd(), `${answers.privKeyPath}.onboard`),
                };

                // POST board
                log('Posting board request');

                let { location, clientUrl } = await postBoard({ server: answers.server });

                // overwrite domain to be the right sso instance
                clientUrl = clientUrl.replace(/(https:\/\/)([\da-z.-]*)/, 'https://' + answers.server);
                location = location.replace(/(https:\/\/)([\da-z.-]*)/, 'https://' + answers.server);

                // body url pour client
                log(`please visit ${clientUrl}`);
                opn(clientUrl);

                const tempCode = new url.URLSearchParams(url.parse(clientUrl).search).get('code');
                const timeout = WAIT_POLL;
                // location -> poll

                log('Polling onboarding result');
                // eslint-disable-next-line no-unused-vars
                const { statusCode, body } = await poll({ location, timeout });
                const userInfo = JSON.parse(body);

                // JSON is like :
                // {
                //     "user": "coucou4@test.mybraincube.com",
                //     "productId": "e3fd395e-2698-4c86-bee9-3c11b88c3341",
                //     "step": "BOARD_DETAILS_AVAILABLE",
                //     "owner": "adrien.deltour@ipleanware.com",
                //     "productName": "test-transfert4"
                // }

                log('Generating Key Pair');
                const keys = await generateKeyPair(externalPrivKeyPath);

                log('Saving Key Pair');
                await saveKeyPair(paths, keys);

                log('Generating Csr');
                const csr = await getCsr(userInfo, keys);

                log('Signing CSR');
                csr.sign(keys.privateKey);

                log('Verifing CSR');
                // eslint-disable-next-line no-unused-vars
                const verified = csr.verify();
                log('Saving CSR');
                const pemCsr = forge.pki.certificationRequestToPem(csr);
                await saveCsr(paths, pemCsr);

                // now post csr
                log('Posting CSR');
                const csrPost = await postCsr({ server: answers.server, csr: pemCsr, code: tempCode });
                log('CSR posted');
                log('saving certificate');
                await saveCert(paths, csrPost.data, keys);

                if (answers.check !== false) {
                    log('Check connection');
                    await checkConnection(answers.server,
                        userInfo.productId,
                        csrPost.data,
                        forge.pki.privateKeyToPem(keys.privateKey),
                        WAIT_PUSH);
                }

                await saveOnboardFile(paths, Object.assign({}, userInfo, paths));

                log(`Private key path is ${paths.privateKey}`);
                log(`Public key path is ${paths.publicKey}`);
                log(`CSR path is ${paths.csr}`);
                log(`Cert path is ${paths.cert}`);
                log(`Onboard file path is ${paths.onboard}`);

            } catch (e) {
                /* eslint-disable no-console */
                console.error(e);
            } finally {
                log('END');
            }
        }
    });
}

export default {};
